### PR TITLE
Add happy-path test cases for Model and ApiClient

### PR DESCRIPTION
**Pull request by natebosch**
Refactor `ApiClient` to go from `Map` to `Map` instead of bytes to
`String`. This makes it easier to write a stub client which is agnostic
to the ordering of values in the maps. Move the responsibilities for
UTF8 and JSON encoding to the `HttpApiClient`.

Eagerly call nested `.toJson()` methods for child fields in `.toJson()`
implementations. This allows deep equality checking for the stub client
without implementing `operator ==` for the data model classes.

Add a `createModelWithClient` method in `src/model.dart` which will not
be exported to the package public API. Use it in the test to create a
model with a stubbed client.

Add tests for each call in `GenerativeModel`. Tests validate
- Encoding from arguments to the JSON schema.
- Building the correct `Uri` for the task and model.
- Parsing the result to the Dart data model.

Add tests for unary and streaming calls in `HttpClient`. Tests validate
- Headers are set correctly.
- Encoding and decoding UTF8 and JSON.
- Using streaming with `alt=sse` query parameter and parsing of `data: `
  lines in the response.

Add Matcher helper methods for classes checked during tests that don't
have `operator ==`.

### CODE REVIEW

File: pkgs/google_generative_ai/lib/src/client.dart [hunk 1]
```diff
@@ -14,42 +14,44 @@
...
 
   @override
-  Future<String> makeRequest(Uri uri, Uint8List body) async {
+  Future<Map<String, Object?>> makeRequest(
+      Uri uri, Map<String, Object?> body) async {
     final response = await http.post(uri,
-        headers: {..._headers, 'Content-Type': 'application/json'}, body: body);
-    return utf8.decode(response.bodyBytes);
+        headers: {..._headers, 'Content-Type': 'application/json'},
+        body: utf8.encode(jsonEncode(body)));
```
Comment (devoncarew) [inline pkgs/google_generative_ai/lib/src/client.dart:43]: for readability?

```suggestion
        body: utf8.encode(jsonEncode(body)),);
```

File: pkgs/google_generative_ai/lib/src/model.dart [hunk 2]
```diff
@@ -42,34 +41,33 @@ final class GenerativeModel {
...
-      {required String model,
-      required String apiKey,
-      List<SafetySetting> safetySettings = const [],
-      GenerationConfig? generationConfig})
+
+  factory GenerativeModel(
+          {required String model,
+          required String apiKey,
+          List<SafetySetting> safetySettings = const [],
+          GenerationConfig? generationConfig}) =>
```
Comment (devoncarew) [inline pkgs/google_generative_ai/lib/src/model.dart:49]:
```suggestion
          GenerationConfig? generationConfig},) =>
```

File: pkgs/google_generative_ai/lib/src/model.dart [hunk 3]
```diff
@@ -134,11 +141,23 @@ final class GenerativeModel {
...
       if (title != null) 'title': title
     };
-    final response = await _makeRequest(Task.embedContent, parameters);
+    final response =
+        await _client.makeRequest(_taskUri(Task.embedContent), parameters);
     return parseEmbedContentReponse(response);
   }
 }
+
+GenerativeModel createModelwithClient(
```
Comment (devoncarew) [inline pkgs/google_generative_ai/lib/src/model.dart:154]: This is only used for testing; I'm not sure if this method is visible outside this library, but assuming it is, perhaps annotate with `visibleForTesting`? Good to document this in either case.

File: pkgs/google_generative_ai/pubspec.yaml [hunk 4]
```diff
@@ -10,10 +10,11 @@ environment:
   sdk: ^3.2.0
 
 dependencies:
-  crypt: ^4.3.0
```
Comment (devoncarew) [inline pkgs/google_generative_ai/pubspec.yaml:13]: +1 - nice to lose a dep

File: pkgs/google_generative_ai/test/generative_model_test.dart [hunk 5]
```diff
@@ -0,0 +1,154 @@
+import 'package:google_generative_ai/google_generative_ai.dart';
```
Comment (devoncarew) [inline pkgs/google_generative_ai/test/generative_model_test.dart:1]: nit: copyright header

File: pkgs/google_generative_ai/test/generative_model_test.dart [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                      'content': {
+                        'role': 'model',
+                        'parts': [
+                          {'text': result}
+                        ]
+                      }
+                    }
+                  ]
+                }
+            ]);
```
Comment (devoncarew) [inline pkgs/google_generative_ai/test/generative_model_test.dart:89]:
```suggestion
            ],);
```

File: pkgs/google_generative_ai/test/http_api_client_test.dart [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import 'dart:convert';
```
Comment (devoncarew) [inline pkgs/google_generative_ai/test/http_api_client_test.dart:1]: copyright header

File: pkgs/google_generative_ai/test/utils/matchers.dart [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import 'package:google_generative_ai/google_generative_ai.dart';
```
Comment (devoncarew) [inline pkgs/google_generative_ai/test/utils/matchers.dart:1]: copyright header

File: pkgs/google_generative_ai/test/utils/stub_client.dart [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import 'package:collection/collection.dart';
```
Comment (devoncarew) [inline pkgs/google_generative_ai/test/utils/stub_client.dart:1]: header

**Review by devoncarew — APPROVED** [review]
lgtm; great refactoring for testability.

Feel free to ignore the tailing comma suggested edits, though I suspect some of them will help w/ idents and readability in the files.

I would comment back in the `dart test` check at the end of the `google_generative_ai.yml` workflow file before landing this.

File: pkgs/google_generative_ai/lib/src/client.dart (hunk 1 above)
Comment (natebosch) [reply pkgs/google_generative_ai/lib/src/client.dart]: Yeah it looks better that way. Hopefully soon we won't need to make these choices manually 🤞

File: pkgs/google_generative_ai/lib/src/model.dart (hunk 2 above)
Comment (natebosch) [reply pkgs/google_generative_ai/lib/src/model.dart]: Done.

File: pkgs/google_generative_ai/lib/src/model.dart (hunk 3 above)
Comment (natebosch) [reply pkgs/google_generative_ai/lib/src/model.dart:158]: I'll add a comment for now. Adding a dependency on `package:meta` is low-cost - but I don't think it's worth for just this annotation.

File: pkgs/google_generative_ai/test/generative_model_test.dart (hunk 5 above)
Comment (natebosch) [reply pkgs/google_generative_ai/test/generative_model_test.dart:15]: Oops. Done.

File: pkgs/google_generative_ai/test/generative_model_test.dart (hunk 6 above)
Comment (natebosch) [reply pkgs/google_generative_ai/test/generative_model_test.dart]: Good call. Added trailing commas for all these subsequent collection arguments.

File: pkgs/google_generative_ai/test/http_api_client_test.dart (hunk 7 above)
Comment (natebosch) [reply pkgs/google_generative_ai/test/http_api_client_test.dart:15]: Done.

File: pkgs/google_generative_ai/test/utils/matchers.dart (hunk 8 above)
Comment (natebosch) [reply pkgs/google_generative_ai/test/utils/matchers.dart:15]: Done.

File: pkgs/google_generative_ai/test/utils/stub_client.dart (hunk 9 above)
Comment (natebosch) [reply pkgs/google_generative_ai/test/utils/stub_client.dart:15]: Done.

**Pull request merged**